File: BreadthFirstSearch.js
class Vertex {
  constructor(value) {
    this.value = value;
    this.isVisited = false;
  }
  setVisit = (isVisited) => {
    this.isVisited = isVisited;
  };
}
class Edge {
  constructor(from, to, weight = 1) {
    this.from = from;
    this.to = to;
    this.weight = weight;
  }
}
class Graph {
  vertexList = [];
  edgeList = [];
  _addVertex = (value) => {
    const vertex = new Vertex(value);
    this.vertexList.push(vertex);
    return vertex;
  };
  addEdgeWithVertexes = (from, to) => {
    const fromVertex = this._addVertex(from);
    const toVertex = this._addVertex(to);
    this.edgeList.push(new Edge(fromVertex, toVertex));
  };
  _setVisitedByVertex = (vertex, isVisited) => {
    this.vertexList.forEach(
      (v) => v.value === vertex.value && v.setVisit(isVisited)
    );
    this.edgeList.forEach(
      (e) =>
        (e.to.value === vertex.value && e.to.setVisit(isVisited)) ||
        (e.from.value === vertex.value && e.from.setVisit(isVisited))
    );
  };
  getVertexesByVertexValue = (vertexValue) => {
    return this.edgeList
      .filter((e) => e.from.value === vertexValue)
      .map((e) => e.to);
  };
  breadthFirstSearch = (startVertexValue, finishVertexValue) => {
    if (startVertexValue === finishVertexValue) {
      this.vertexList.forEach((v) => this._setVisitedByVertex(v, false));
      return true;
    }
    const vertexes = this.getVertexesByVertexValue(startVertexValue).filter(
      (v) => !v.isVisited
    );
    for (const key in vertexes) {
      this._setVisitedByVertex(vertexes[key], true);
      const isExact = this.breadthFirstSearch(
        vertexes[key].value,
        finishVertexValue
      );
      if (isExact) {
        return true;
      }
    }
    return false;
  };
}

const graph = new Graph();
graph.addEdgeWithVertexes(1, 2);
graph.addEdgeWithVertexes(1, 3);
graph.addEdgeWithVertexes(1, 4);
graph.addEdgeWithVertexes(2, 3);
graph.addEdgeWithVertexes(2, 4);
graph.addEdgeWithVertexes(3, 4);
graph.addEdgeWithVertexes(4, 5);
graph.addEdgeWithVertexes(3, 5);
graph.addEdgeWithVertexes(5, 4);
graph.addEdgeWithVertexes(4, 3);
graph.addEdgeWithVertexes(3, 2);
graph.addEdgeWithVertexes(2, 1);
console.log(graph.breadthFirstSearch(1, 5), graph.breadthFirstSearch(1, 6));

File: BreadthFirstSearch.py
class Vertex():
    def __init__(self, name):
        self.name = name

    def __str__(self):
        return self.name
    visite = 1


class Edge():
    def __init__(self, From, To, weight=1):
        self.From = From
        self.To = To
        self.weight = weight

    def __str__(self):
        return ("From:"+str(self.From)+" To:"+str(self.To))


class Graph():
    VertexList, EdgeList = [], []

    def AddVertex(self, vertex):
        self.VertexList.append(Vertex(vertex))

    def AddEdge(self, From, To, width=1):
        self.EdgeList.append(Edge(From, To, width))

    def GetVertex(self, vertex):
        result = []
        # Go through array and check who connected with current vertex
        for a in self.EdgeList:
            if a.From.name == vertex.name:
                result.append(a.To)
        return result

    def BreadthFirstSearch(self, start, finish):
        # Recursion Key
        if start.name == finish.name:
            print("True)")
        curVertex = self.GetVertex(start)
        for a in curVertex:
            # Do recursion if vertex dont visited
            if a.visite == 1:
                a.visite = 0
                self.BreadthFirstSearch(a, finish)


# Realisation Graph
graph = Graph()
for a in range(1, 8):
    graph.AddVertex(a)
graph.AddEdge(Vertex(1), Vertex(2))
graph.AddEdge(Vertex(1), Vertex(3))
graph.AddEdge(Vertex(2), Vertex(4))
graph.AddEdge(Vertex(4), Vertex(5))
graph.AddEdge(Vertex(4), Vertex(6))
graph.AddEdge(Vertex(6), Vertex(4))
graph.AddEdge(Vertex(5), Vertex(7))
graph.BreadthFirstSearch(Vertex(4), Vertex(7))
